src/Data/Singletons/TypeLits.hs

```diff
@@ -98,32 +100,63 @@ no_term_level_syms = error "The kind `Symbol` may not be used at the term level.
 $(genDefunSymbols [''KnownNat, ''KnownSymbol])
 
 ------------------------------------------------------------
--- Div, Mod, DivMod, and friends
+-- Log2, Div, Mod, DivMod, and friends
 ------------------------------------------------------------
 
+{- | Adapted from GHC's source code.
+
+Compute the logarithm of a number in the given base, rounded down to the
+closest integer. -}
+genLog2 :: Natural -> Natural
+genLog2 x = exactLoop 0 x
+  where
+  exactLoop s i
+    | i == 1     = s
+    | i < 2      = s
+    | otherwise  =
+        let s1 = s + 1
+        in s1 `seq` case divMod i 2 of
+                      (j,r)
+                        | r == 0    -> exactLoop s1 j
+                        | otherwise -> underLoop s1 j
+
+  underLoop s i
+    | i < 2  = s
+    | otherwise = let s1 = s + 1 in s1 `seq` underLoop s1 (div i 2)
+
+
+sLog2 :: Sing x -> Sing (TN.Log2 x)
+sLog2 sx =
+    let x   = fromSing sx
+    in case x of
+         0 -> error "log2 of 0"
+         _ -> case TN.someNatVal (genLog2 x) of
+                SomeNat (_ :: Proxy res) -> unsafeCoerce (SNat :: Sing res)
+$(genDefunSymbols [''TN.Log2])
+
+sDiv :: Sing x -> Sing y -> Sing (Div x y)
+sDiv sx sy =
+    let x   = fromSing sx
+        y   = fromSing sy
+        res = TN.someNatVal (x `div` y)
+    in case res of
+         SomeNat (_ :: Proxy res) -> unsafeCoerce (SNat :: Sing res)
+infixl 7 `sDiv`
+$(genDefunSymbols [''Div])
+
+sMod :: Sing x -> Sing y -> Sing (Mod x y)
+sMod sx sy =
+    let x   = fromSing sx
+        y   = fromSing sy
+        res = TN.someNatVal (x `mod` y)
+    in case res of
+         SomeNat (_ :: Proxy res) -> unsafeCoerce (SNat :: Sing res)
+infixl 7 `sMod`
+$(genDefunSymbols [''Mod])
+
 $(promoteOnly [d|
-  -- https://ghc.haskell.org/trac/ghc/ticket/13652 asks for these in GHC.TypeLits.
-  -- That would be nice, since this implementation is horribly slow.
   divMod :: Nat -> Nat -> (Nat, Nat)
-  divMod _ 0 = error "Division by zero."
-  divMod x y =
-    let (d, m) = (divMod' x (y-1) 0 (y-1))
-    in (d, (y-1) - m)
-
-  divMod' :: Nat -> Nat -> Nat -> Nat -> (Nat, Nat)
-  divMod' 0 _ q u = (q, u)
-  divMod' n y q 0 = divMod' (n-1) y (q+1) y
-  divMod' n y q u = divMod' (n-1) y q     (u-1)
-
-  div :: Nat -> Nat -> Nat
-  div _ 0 = error "Division by zero."
-  div x y = fst (divMod x y)
-  infixl 7 `div`
```

Comment:
One question you might ask is: do the built-in `Div` and `Mod` type families from `GHC.TypeNats` have the same fixity (`infixl 7`)? Sadly, the answer is currently "no", although I'm seeking to correct this discrepancy in [GHC Trac #14640](https://ghc.haskell.org/trac/ghc/ticket/14640).